### PR TITLE
feat: Make teleport work with s3 and duckdb

### PR DESCRIPTION
**Pull request by mederka**
Teleport + S3 + DuckDB

Example profiles.yml:

```
jaffle_shop:
  target: dev_with_fal
  outputs:
    dev_with_fal:
      type: duckdb
      path: "duck_db_dbt_dump.db"
      python_adapter:
        type: fal
        teleport:
          type: s3
          s3_bucket: fal-dbt-test
          s3_region: my_region
          s3_access_key_id: my_key_id
          s3_access_key: my_key
```

### CODE REVIEW

File: adapter/src/dbt/adapters/fal/teleport_support/duckdb.py [hunk 1]
```diff
@@ -23,17 +25,35 @@ def teleport_from_external_storage(self, relation: BaseRelation, relation_path:
         url = teleport_info.build_url(relation_path)
 
         with self._db_adapter.connection_named('teleport:copy_from'):
-            rendered_macro = self._db_adapter.execute_macro('duckdb__copy_from', kwargs={'relation': relation, 'url': url})
+            if self.credentials.type == TeleportTypeEnum.REMOTE_S3:
+                # Putting this in __init__ didn't work, looks like it has to be done with each new connection
```
Comment (chamini2) [inline adapter/src/dbt/adapters/fal/teleport_support/duckdb.py:29]: That is true! It has to be setup on each

File: adapter/src/dbt/adapters/fal/teleport_support/duckdb.py [hunk 2]
```diff
@@ -23,17 +25,35 @@ def teleport_from_external_storage(self, relation: BaseRelation, relation_path:
...
         with self._db_adapter.connection_named('teleport:copy_to'):
+            if self.credentials.type == TeleportTypeEnum.REMOTE_S3:
+                self._setup_s3()
             rendered_macro = self._db_adapter.execute_macro('duckdb__copy_to', kwargs={'relation': relation, 'url': url})
             self._db_adapter.execute(rendered_macro)
 
         return rel_path
+
+    def _setup_s3(self):
+        self._db_adapter.execute("INSTALL httpfs")
```
Comment (chamini2) [inline adapter/src/dbt/adapters/fal/teleport_support/duckdb.py:55]: Can we do `INSTALL` on init? Maybe next to `self._db_adapter.execute("INSTALL 'parquet'")`

File: adapter/src/dbt/adapters/fal/teleport_support/duckdb.py (hunk 1 above)
Comment (chamini2) [reply adapter/src/dbt/adapters/fal/teleport_support/duckdb.py:29]: Btw, does that mean we should be doing `LOAD 'parquet'` on each connection?

Comment (chamini2) [reply adapter/src/dbt/adapters/fal/teleport_support/duckdb.py:30]: For me it failed before. I will test locally.

File: adapter/src/dbt/include/fal/macros/teleport_duckdb.sql [hunk 3]
```diff
@@ -6,3 +6,8 @@
     CREATE OR REPLACE TABLE {{ relation }} AS
         SELECT * FROM '{{ url }}'
 {%- endmacro %}
+
+{% macro duckdb__copy_from_s3(relation, url) -%}
+    CREATE OR REPLACE TABLE {{ relation }} AS
+        SELECT * FROM read_parquet('{{ url }}');
+{%- endmacro %}
```
Comment (chamini2) [inline adapter/src/dbt/include/fal/macros/teleport_duckdb.sql:13]: I think we could change this to `duckdb__copy_from_parquet` and use it for both local and S3, and delete the `duckdb__copy_from` one.

**Comment by chamini2** [comment]
testing locally I got: 

```
14:53:54  7 of 8 START python table model main.orders_forecast ........................... [RUN]
14:53:58  Unhandled error while executing model.jaffle_shop.orders_forecast
Missing optional dependency 'fsspec'.  Use pip or conda to install fsspec.
14:53:58  7 of 8 ERROR creating python table model main.orders_forecast .................. [ERROR in 4.02s]
14:53:58  8 of 8 SKIP relation main.orders_forecast_filter ............................... [SKIP]
14:53:58
14:53:58  Finished running 3 view models, 5 table models in 0 hours 0 minutes and 4.18 seconds (4.18s).
14:53:58
14:53:58  Completed with 1 error and 0 warnings:
14:53:58
14:53:58  Missing optional dependency 'fsspec'.  Use pip or conda to install fsspec.
```

File: adapter/src/dbt/adapters/fal/teleport_support/duckdb.py (hunk 1 above)
Comment (chamini2) [reply adapter/src/dbt/adapters/fal/teleport_support/duckdb.py:30]:
```
❯ duckdb ~/duck_db_dbt_dump.db
v0.5.1 7c111322d
Enter ".help" for usage hints.
D load httpfs;
D SET s3_region='x';
D SET s3_access_key_id='x';
D SET s3_secret_access_key='x';
D SELECT * FROM read_parquet('s3://mybucket/teleport/main.main.orders_forecast.parquet');
Error: Catalog Error: Function with name read_parquet is not on the catalog, but it exists in the parquet extension. To Install and Load the extension, run: INSTALL parquet; LOAD parquet;
```

Comment (chamini2) [reply adapter/src/dbt/adapters/fal/teleport_support/duckdb.py:30]: But not putting it the code it still works. I am assuming it is loaded by dbt-duckdb somewhere... But can we still include it to be safe?

**Comment by mederka** [comment]
> testing locally I got:
> 
> ```
> 14:53:54  7 of 8 START python table model main.orders_forecast ........................... [RUN]
> 14:53:58  Unhandled error while executing model.jaffle_shop.orders_forecast
> Missing optional dependency 'fsspec'.  Use pip or conda to install fsspec.
> 14:53:58  7 of 8 ERROR creating python table model main.orders_forecast .................. [ERROR in 4.02s]
> 14:53:58  8 of 8 SKIP relation main.orders_forecast_filter ............................... [SKIP]
> 14:53:58
> 14:53:58  Finished running 3 view models, 5 table models in 0 hours 0 minutes and 4.18 seconds (4.18s).
> 14:53:58
> 14:53:58  Completed with 1 error and 0 warnings:
> 14:53:58
> 14:53:58  Missing optional dependency 'fsspec'.  Use pip or conda to install fsspec.
> ```
> 
> And after `pip install fsspec` I got
> 
> ```
>   File "/Users/matteo/.pyenv/versions/3.8.13/envs/dbt38-dd/lib/python3.8/site-packages/fsspec/registry.py", line 236, in get_filesystem_class
>     raise ImportError(bit["err"]) from e
> ImportError: Install s3fs to access S3
> ```
> 
> `pip install fsspec[s3]` fixed it
> 
> We could do an "extra" group for teleport-s3 / teleport-local / teleport-* dependencies
> 
> `pip install fal-dbt[teleport-s3]`

`s3fs` is already in `pyproject.toml`. I think we should try to fix the dependencies so we can have a lockfile.

**Pull request merged**